028-service-binding.md

```diff
@@ -0,0 +1,340 @@
+2. Application uses multiple `ServiceBinding` resources to get information
+3. Application binds to the `KafkaUser` resource which contains all required information
+4. Application can bind to a single custom resource to get all information
+
+### 1 - Application binds to KafkaUser only
+
+In this approach the application creates a `ServiceBinding` for a specific `KafkaUser` CR. The credential information is bound to the application and the cluster certificate and bootstrap server address is obtained by the application in another way.
+
+Required changes:
+
+ - Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+ - Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+
+Considerations:
+
+ - In this approach a lot of work is still left up to the user to discover the cluster certificate and the bootstrap server address to use.
+
+### 2 - Application uses multiple ServiceBinding resources
+
+In this approach the application creates two `ServiceBinding` resources, one that will bind to the `Kafka` CR and one that will bind to the `KafkaUser` CR.
+
+Required changes:
+
+- Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+- Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+- Update the `Kafka` CR to include a `status.binding` field, where the `name` refers to a new secret containing both the cluster certificate and bootstrapServers
+- Add a new secret that contains the bootstrap servers for each listener and the cluster certificate information
+
+Considerations:
+
+ - This approach results in a new secret being required that contains copies of existing secrets
+ - It does not help the user to decide which listener to use
+
+### 3 - Application binds to KafkaUser resource that contains all details
+
+In this approach the application only binds to the `KafkaUser` and the secret created for the `KafkaUser` contains all the required details.
+
+Required changes:
+
+- Update the `KafkaUser` CR to include a `status.binding` field where the `name` matches the current field `secret` in the status
+- Update the secret created for the `KafkaUser` to include the username when the `KafkaUser` is using type SASL SCRAM
+- Update the secret created for the `KafkaUser` to include the bootstrap servers for each listener
+- Update the secret created for the `KafkaUser` to include the cluster certificate information
+
+Considerations:
+
+- This approach results in a lot more information being added to each `KafkaUser` secret
+- It does not help the user to decide which listener to use
+
+Possible extensions:
+
+ - Could only add bootstrap server addresses for listeners that support the authentication type that the `KafkaUser` supports
```

Comment:
This kind of validation is good. Using the `spec.binding.listener` idea failure could be a reason in an unready condition.